### ISSUE TITLE
systemd-resolved: allow configurable bind address

### ISSUE THREAD

**Issue by bacoboy** [systemd/systemd#10859]
**Is your feature request related to a problem? Please describe.**
On servers which run containers such as docker and via `runc`, they cannot make use of the local caching resolver because it is bound to a non-reachable IP address.
The current solution has this [hard-coded with no option to override](https://github.com/systemd/systemd/blob/4d09e1c8bab1d684172b1f277f3213825b30d2d9/src/resolve/resolved-dns-stub.c#L425-L427) what seems like a sane default.

**Describe the solution you'd like**
I'd like an environment variable to set the `BIND_ADDRESS` which, if specified, will be used instead of the hard coded `127.0.0.53` IP.  In this way, I can decide to open up the defaults (i.e. `0.0.0.0` or the server's primary interface's IP)

**Describe alternatives you've considered**
The only alternative to providing "on server" DNS caching for docker/containers is to:
* run a second resolver such as dnsmasq -- but that seems silly to have 2
* recompile/fork the code to remove the hard-coded value
* overriding iptables for the default docker bridge network -- `sudo iptables -t nat -A PREROUTING -p udp -s 172.17.0.0/16 --dport 53 -i docker0 -j DNAT --to-destination 127.0.0.53:53` - but `SO_BINDTODEVICE` prevents responding to these packets.
* ask for a configurable override (this ticket)

**Comment by b1tninja** [comment]
I'll attempt to implement and make a PR

**Comment by b1tninja** [comment]
https://github.com/systemd/systemd/compare/master...b1tninja:dns_bind_addr

Would appreciate feedback, still need to wire up config parsing, but i've moved it into the Manager struct compiled and it listens, so thats a good sign

**Comment by yrro** [comment]
I've been using the following workaround:

$ cat /etc/systemd/system/resolved-proxy\@.service 
[Unit]
Description=DNS proxy for %I (launched via /etc/udev/rules.d/95-resolved-proxy.rules)

```
$ cat /etc/systemd/system/resolved-proxy@.service
[Service]
Type=simple
User=dnsmasq
ExecStart=/usr/sbin/dnsmasq --no-daemon --no-hosts --interface=%I --no-dhcp-interface=%I --except-interface=lo --bind-interfaces
ProtectSystem=full
NoNewPrivileges=yes
AmbientCapabilities=CAP_NET_BIND_SERVICE

$ cat /etc/udev/rules.d/95-resolved-proxy.rules 
ACTION=="add", SUBSYSTEM=="net", NAME="docker0", ENV{SYSTEMD_WANTS}="resolved-proxy@$name"
```

**Comment by b1tninja** [comment]
I made some progress on this awhile back but didn't think it was a good implementation. Specifically the structures that handle the address, and representing them as strings. I switched to using networkd for dhcp server so it isn't much of an issue for me anymore but I would like to bind to in addr any still, so I can have it be the resolver for my network, instead of forwarding to lo.

If anyone can offer some feedback maybe I'll resume, but doesn't feel PR worthy in it's current form
